fms_extras/utils/generation.py

```diff
@@ -3,19 +3,564 @@
+
+    return result_input_ids, unflat_indices, flat_indices
+
+
+def __free_incorrect_tokens_and_sequences(
+    kv_cache_manager: PagedKVCacheManager,
+    child_sequence_ids_list: List[List[int]],
+    best_guess: torch.Tensor,
+    n_correct: torch.Tensor,
+    decode_seq_length: int,
+) -> List[int]:
+    """
+    Free all non-best candidate sequences from the kv-cache and for best candidate
+    sequences, logically remove tokens based on the number of incorrect tokens
+    generated
+
+    Args:
+        kv_cache_manager: PagedKVCacheManager
+            the paged attention kv-cache manager
+        child_sequence_ids_list: List[List[int]]
+            the list of child sequences per parent
+        best_guess: torch.Tensor
+            a tensor containing the best candidate per sequence in the batch
+        n_correct: torch.Tensor
+            a tensor containing the number of correct tokens best candidate in the
+            batch
+        decode_seq_length: int
+            the model input length at decode step
+
+    Returns:
+    List[int]
+        a list of the new parent sequence ids for next decode step
+    """
+    parent_sequence_ids = []
+    for parent_index, child_sequence_ids in enumerate(child_sequence_ids_list):
+        best_index = int(best_guess[parent_index].item())
+
+        # free all bad candidates
+        kv_cache_manager.free_sequences(
+            child_sequence_ids[:best_index] + child_sequence_ids[best_index + 1 :]
+        )
+
+        # decrease the context length of the sequence which used to be sequence length + inp_len by the number of incorrect tokens
+        # for the best candidate
+        best_sequence_id = child_sequence_ids[best_index]
+        parent_sequence_ids.append(best_sequence_id)
+        kv_cache_manager.remove_tokens(
+            best_sequence_id, decode_seq_length - n_correct[parent_index].item() - 1  # type: ignore
+        )
+    return parent_sequence_ids
+
+
+def __perform_correctness_checks(
+    input_ids: torch.Tensor,
+    next_vals: torch.Tensor,
+    embeds: torch.Tensor,
+) -> Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+    """
+    Check for the correctness of speculator predictions and get the indices of the
+    best candidates, the number of tokens correct, and the base model output values
+    for that candidate (tokens and embeddings)
+
+    Args:
+        input_ids: torch.Tensor
+            the original input ids unflattened
+        next_vals: torch.Tensor
+            a tensor of the next tokens
+        embeds: torch.Tensor
+            the output embeddings for the best guesses
+
+    Returns:
+    Tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]
+        a tensor of the next tokens per best guess, a tensor of the embeds per best
+        guess, a tensor of the number of correct tokens per best guess, and a tensor
+        containing the best guess amongst all candidates per sequence
```

Comment:
it might help to add more detail on each of the return types, dimensions, and contents.